### PR TITLE
Wishlist: frontend part + backend fixes

### PR DESCRIPTION
**Pull request by iryb**
Task https://app.asana.com/0/1202525679663009/1202720801120417

- Connected backend logic to the frontend
- Added Wishlist component
- Replaced hardcoded data with props in Car card component

chrome-capture-2022-7-19 (1)

### CODE REVIEW

File: backend/src/controllers/preferences/wishlist.controller.ts [hunk 1]
```diff
@@ -33,21 +33,16 @@ const setWishlist = async (
 };
 
 const deleteWishlist = async (
-  req: TypedRequestQuery<WishlistInput>,
```
Comment (oleksandr-danylchenko) [inline backend/src/controllers/preferences/wishlist.controller.ts:36]: Why did you remove the `WishlistInput` type? :)
 
I would expect it to be declared as the `type WishlistInput = { wishlistId: string; }`

File: frontend/src/components/common/icons/heart/heart.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { SvgIcon } from '@mui/material';
+
+export const Heart = (props: JSX.IntrinsicAttributes): React.ReactElement => (
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/common/icons/heart/heart.tsx]: It's better to name it more explicitly - `HeartIcon`

File: frontend/src/common/types/card/card.ts [hunk 3]
```diff
@@ -12,4 +18,17 @@ interface CarDataListPropsType {
...
 }
 
-export type { CarDataType, CarDataPropsType, CarDataListPropsType };
+interface ExtendedCarCardPropsType {
+  type: 'model' | 'complectation';
+  car: ComplectationResponseDto | ModelResponseDto;
+  deleteWishlist: (wishlistId: string) => Promise<void>;
+  createWishlist: (args: WishlistInput) => Promise<void>;
+  isLiked: boolean;
+}
```
Comment (oleksandr-danylchenko) [inline frontend/src/common/types/card/card.ts]:
```suggestion
interface ExtendedCarCardPropsType {
  type: 'model' | 'complectation';
  car: ComplectationResponseDto | ModelResponseDto;
  createWishlist: (args: WishlistInput) => Promise<void>;
  deleteWishlist: (wishlistId: string) => Promise<void>;
  isLiked: boolean;
}
```

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 4]
```diff
@@ -1,35 +1,100 @@
...
+  const formatPriceNumber = (num: number): string => {
+    const price = num.toString().split('.');
+    let numberPart = price[0];
+    const floatPart = price.length > 1 ? '.' + price[1] : '';
+    const rgx = /(\d+)(\d{3})/;
+    while (rgx.test(numberPart)) {
+      numberPart = numberPart.replace(rgx, '$1' + ' ' + '$2');
+    }
+    return numberPart + floatPart;
+  };
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: This needs to be outside of the React component, in helpers/utils. Because this function definitely can be reused. Also, try to minimize the creation of the functions inside the react components because they are getting recreated on each rerender

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 5]
```diff
@@ -1,35 +1,100 @@
...
+const NewCarCard: React.FC<ExtendedCarCardPropsType> = (props) => {
+  const formatPriceNumber = (num: number): string => {
+    const price = num.toString().split('.');
+    let numberPart = price[0];
+    const floatPart = price.length > 1 ? '.' + price[1] : '';
+    const rgx = /(\d+)(\d{3})/;
+    while (rgx.test(numberPart)) {
+      numberPart = numberPart.replace(rgx, '$1' + ' ' + '$2');
+    }
+    return numberPart + floatPart;
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: Could you explain what happens here? Why is it so complicated?)
Doesn't the `.toFixed` to `.toLocaleString({ currency: ...})` fit your needs?

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 6]
```diff
@@ -1,35 +1,100 @@
...
+  };
+
+  const minPrices = props.car.pricesRanges.map(
+    (price: { price_start: number; price_end: number }) => price.price_start,
+  );
+  const minPrice = formatPriceNumber(Math.min.apply(null, minPrices));
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: Please use the spread operator instead of the `.apply`

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 7]
```diff
@@ -1,35 +1,100 @@
...
+
+  const handleCreate = (): void => {
+    const data: WishlistInput = {
+      modelId: props.type === 'model' ? props.car.id : undefined,
+      complectationId:
+        props.type === 'complectation' ? props.car.id : undefined,
+    };
+
+    props.createWishlist(data);
+  };
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: How about such simplification? :slightly_smiling_face: 

```suggestion
  const handleCreate = (): void => {
   const { type, car: { id: carId }, createWishlist } = props;
    
    const data: WishlistInput = type === 'model' ? { modelId: carId  } : { complectationId: carId };
    createWishlist(data);
  };
```

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 8]
```diff
@@ -1,35 +1,100 @@
...
+  const handleLikeClick = (event: React.MouseEvent): void => {
+    event.stopPropagation();
+    props.isLiked ? handleDelete() : handleCreate();
+  };
+
+  let name = props.car.brand?.name.concat(' ', props.car.name);
+
+  if (props.car.complectationName) {
+    name = props.car.brand?.name.concat(' ', props.car.complectationName);
+  }
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: Please use the template string here. Usage of the `contact` isn't recommended in the modern JS

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 9]
```diff
@@ -1,35 +1,100 @@
...
-import new_car_1 from '@assets/images/mock_car_picture.png';
+import { WishlistInput } from '@autoline/shared/common/types/types';
+import { ExtendedCarCardPropsType } from '@common/types/types';
+import { Heart } from '@components/common/icons/icons';
+import { clsx } from 'clsx';
 
 import styles from './styles.module.scss';
 
-const NewCarCard: FC = () => {
+const NewCarCard: React.FC<ExtendedCarCardPropsType> = (props) => {
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx:12]: Please use props destructuring at the start of the component: `const {...} = props`.
It'll help to avoid contestant `props.` repetition

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 10]
```diff
@@ -1,35 +1,100 @@
...
       <div className={styles.carTitle}>
-        <img className={styles.carLogo} src={bmw} alt="car logo" />
-        <span className={styles.carName}>BMW X5</span>
+        {props.car.brand && (
+          <img
+            className={styles.carLogo}
+            src={props.car.brand?.logoUrl}
+            alt={props.car.brand?.name}
+          />
+        )}
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: A car will always have the `brand`. Why we should use the condition here?

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 11]
```diff
@@ -1,35 +1,100 @@
...
         <img className={styles.button} src={compare} alt="compare button" />
       </div>
-      <img src={new_car_1} alt="car image" className={styles.carImage}></img>
+      {props.car.photoUrls && (
+        <img
+          src={props.car.photoUrls[0]}
+          alt="car image"
+          className={styles.carImage}
+        ></img>
+      )}
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: `photoUrls` is mandatory. Why we use such conditional check?

File: frontend/src/components/new-car-card/new-car-card.tsx [hunk 12]
```diff
@@ -1,35 +1,100 @@
...
+        </button>
         <img className={styles.button} src={compare} alt="compare button" />
       </div>
-      <img src={new_car_1} alt="car image" className={styles.carImage}></img>
+      {props.car.photoUrls && (
+        <img
+          src={props.car.photoUrls[0]}
+          alt="car image"
+          className={styles.carImage}
+        ></img>
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/new-car-card/new-car-card.tsx]: Use React's self-closed tags if they are empty:
```suggestion
        <img
          src={props.car.photoUrls[0]}
          alt="car image"
          className={styles.carImage}
        />
```

File: frontend/src/components/wishlist/wishlist.tsx [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import { Title } from '@components/common/title/title';
+import { NewCarCard } from '@components/new-car-card/new-car-card';
+import { Grid } from '@mui/material';
+import {
+  useGetWishlistsQuery,
+  useCreateWishlistMutation,
+  useDeleteWishlistMutation,
+} from '@store/queries/preferences/wishlist';
+
+const Wishlist = (): JSX.Element => {
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/wishlist/wishlist.tsx]: Please use React's `FC` type

File: frontend/src/components/wishlist/wishlist.tsx [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+              data.models &&
+              data.models.map((model: ModelResponseDto) => {
+                return (
+                  <Grid item xs={12} md={4} key={model.id}>
+                    <NewCarCard
+                      car={model}
+                      type="model"
+                      deleteWishlist={handleDelete}
+                      createWishlist={handleCreate}
+                      isLiked={true}
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/wishlist/wishlist.tsx]:
```suggestion
                    <NewCarCard
                      car={model}
                      type="model"
                      createWishlist={handleCreate}
                      deleteWishlist={handleDelete}
                      isLiked={true}
```

File: frontend/src/components/wishlist/wishlist.tsx [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+                  </Grid>
+                );
+              })}
+          </Grid>
+          <Title element="h4">LIKED COMPLECTATIONS</Title>
+          <Grid container spacing={2}>
+            {data &&
+              data.complectations &&
+              data.complectations.map(
+                (complectation: ComplectationResponseDto) => {
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/wishlist/wishlist.tsx]: Optional chaining can be used here:
```suggestion
            {data?.complectations?.map(
                (complectation: ComplectationResponseDto) => {
```

File: frontend/src/store/queries/preferences/wishlist/index.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        url: `${API.WISHLIST}`,
+        method: 'POST',
+        params,
+      }),
+      invalidatesTags: ['WishlistCars'],
+    }),
+    deleteWishlist: build.mutation<WishlistsResponseDto, string>({
+      query: (wishlistId: string) => ({
+        url: `${API.WISHLIST}?wishlistId=${wishlistId}`,
+        method: 'DELETE',
```
Comment (oleksandr-danylchenko) [inline frontend/src/store/queries/preferences/wishlist/index.ts]: We mustn't construct url params manually because RTKQ already has the `params` prop

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]
It's getting closer :+1: 
But some polishing is required to pass the review :slightly_smiling_face:

File: frontend/src/common/types/card/card.ts (hunk 3 above)
Comment (oleksandr-danylchenko) [reply frontend/src/common/types/card/card.ts]: It's better to follow the CRUD abbr order in props definition

**Comment by oleksandr-danylchenko** [comment]
image
We shouldn't show "Liked complectations" if no complications are liked

File: frontend/src/components/wishlist/wishlist.tsx [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+  };
+
+  const handleDelete = async (args: DeleteWishlistInput): Promise<void> => {
+    await deleteWishlist(args);
+  };
+
+  return (
+    <>
+      {!isLoading && (
+        <>
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/wishlist/wishlist.tsx]: This can be simplified as:
```suggestion
  if (isLoading) return null;
  return (
    <>
```

File: frontend/src/helpers/utils/format-price.ts [hunk 18]
```diff
@@ -0,0 +1,11 @@
+const formatPrice = (num: number): string => {
+  return num
+    .toLocaleString('us-US', {
+      style: 'currency',
+      currency: 'USD',
+      maximumFractionDigits: 0,
+    })
+    .replace(/,/g, ' ');
+};
```
Comment (oleksandr-danylchenko) [inline frontend/src/helpers/utils/format-price.ts]: One-liner
```suggestion
const formatPrice = (num: number): string =>
  num
    .toLocaleString('us-US', {
      style: 'currency',
      currency: 'USD',
      maximumFractionDigits: 0,
    })
    .replace(/,/g, ' ');
```

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]
Minor code style changes + visual changes

**Comment by oleksandr-danylchenko** [comment]
What's going on with the price? :eyes: 
image

**Comment by iryb** [comment]
> image
 We shouldn't show "Liked complectations" if no complications are liked

Resolved in PR #61

**Review by oleksandr-danylchenko — APPROVED** [review]
